parsers/cxx/cxx_parser.c

```diff
@@ -1617,76 +1622,114 @@ bool cxxParserParseAccessSpecifier(void)
+	if(cxxTokenTypeIsOneOf(
+			g_cxx.pToken,
+			CXXTokenTypeEOF | CXXTokenTypeSemicolon | CXXTokenTypeOpeningBracket
+		))
 	{
-		CXX_DEBUG_LEAVE_TEXT("Found EOF/semicolon while parsing if/for/while/switch");
+		CXX_DEBUG_LEAVE_TEXT(
+				"Found EOF/semicolon/opening bracket while parsing if/for/while/switch/catch"
+			);
 		return true;
 	}
 
-	if(cxxTokenTypeIs(g_cxx.pToken,CXXTokenTypeParenthesisChain))
-	{
-		// Extract variables from the parenthesis chain
-		// We handle only simple cases.
-		CXXTokenChain * pChain = g_cxx.pToken->pChain;
+	CXX_DEBUG_ASSERT(
+			cxxTokenTypeIs(g_cxx.pToken,CXXTokenTypeParenthesisChain),
+			"Expected a parenthesis chain here"
+		);
+	
+	CXX_DEBUG_PRINT("Found if/for/while/switch/catch parenthesis chain");
 
-		CXX_DEBUG_ASSERT(
-				pChain->iCount >= 2,
-				"The parenthesis chain must have initial and final parenthesis"
-			);
+	// Extract variables from the parenthesis chain
 
-		// Simple check for cases like if(a & b), if(a * b).
-		// If there is &, && or * then we expect there to be also a = or a ;.
-		if(
-				// & && * not present
-				!cxxTokenChainFirstTokenOfType(
+	CXXTokenChain * pChain = g_cxx.pToken->pChain;
+
+	CXX_DEBUG_ASSERT(
+			pChain->iCount >= 2,
+			"The parenthesis chain must have initial and final parenthesis"
+		);
+
+	// There are several constructs that can fool the parser here.
+	// The most notable ones are:
+	//     if(a & b)
+	//     if(a * b)
+	//     if(a && b)
+	// Too bad that these constructs are also used to declare variables.
+
+	// catch() always contains variable declarations
+	bool bOkToExtractVariables = eKeyword == CXXKeywordCATCH;
+
+	if(!bOkToExtractVariables)
+	{
+		// Parenthesis contents start with a keyword. Things like:
+		//   if(const std::exception & e)
+		//   if(int i ...
+		bOkToExtractVariables = cxxTokenTypeIs(
+				cxxTokenChainAt(pChain,1),
+				CXXTokenTypeKeyword
+			);
+		
+		if(!bOkToExtractVariables)
+		{
+			// If there is &, && or * then we expect there to be also a = or
+			// a semicolon that comes after it.
+			// This is not 100% foolproof but works most of the times.
+	
+			CXXToken * pAndOrStar = cxxTokenChainFirstTokenOfType(
 						pChain,
 						CXXTokenTypeAnd | CXXTokenTypeMultipleAnds |
 						CXXTokenTypeStar
-					) ||
-				// or [=;] present
-				cxxTokenChainFirstTokenOfType(
-						pChain,
+					);
+			
+			if(!pAndOrStar)
+			{
+				bOkToExtractVariables = true;
+			} else {
+				bOkToExtractVariables = cxxTokenChainNextTokenOfType(
+						pAndOrStar,
 						CXXTokenTypeAssignment | CXXTokenTypeSemicolon
-					)
-			)
-		{
-			// Kill the initial parenthesis
-			cxxTokenChainDestroyFirst(pChain);
-			// Fake the final semicolon
-			CXXToken * t = cxxTokenChainLast(pChain);
-			t->eType = CXXTokenTypeSemicolon;
-			vStringClear(t->pszWord);
-			vStringPut(t->pszWord,';');
-
-			// and extract variable declarations if possible
-			cxxParserExtractVariableDeclarations(pChain,0);
+					);
```

Comment:
Visual Studio makes a warning on this line (https://ci.appveyor.com/project/universalctags/ctags/build/1.0.763/job/wim07v3kr02a88r6). How about using trigraph like:
```
bOkToExtractVariables = cxxTokenChainNextTokenOfType(
						pAndOrStar,
 						CXXTokenTypeAssignment | CXXTokenTypeSemicolon
					)? true: false;
```

or ...
```
} else if ( cxxTokenChainNextTokenOfType(
					pAndOrStar,
 						CXXTokenTypeAssignment | CXXTokenTypeSemicolon)) {
            bOkToExtractVariables = true;
}

Reply:
done